### PR TITLE
Flesh out the API docs

### PR DESCRIPTION
**Pull request by SamWilsn**

### CODE REVIEW

File: etk-analyze/Cargo.toml [hunk 1]
```diff
@@ -1,18 +1,24 @@
 [package]
 name = "etk-analyze"
 version = "0.1.0"
-authors = ["Sam Wilson <sam.wilson@mesh.xyz>"]
+authors = ["Sam Wilson <sam.wilson@mesh.xyz>", "lightclient <lightclient@protonmail.com>"]
 license = "MIT OR Apache-2.0"
 edition = "2018"
+description = "Ethereum Toolkit analysis tools"
```
Comment (lightclient) [inline etk-analyze/Cargo.toml:7]: I think "tool kit" is actually correct: https://www.copyeditcat.com/2014/05/08/editors-tool-kit/#:~:text=Sure%2C%20toolkit%20looks%20correct%2C%20but%20maybe%20it's%20really%20two%20words.&text=According%20to%20the%20Canadian%20Oxford,indeed%20two%20words%3A%20tool%20kit.

File: etk-asm/src/asm.rs [hunk 2]
```diff
@@ -113,10 +174,13 @@ impl Default for Assembler {
 }
 
 impl Assembler {
+    /// Create a new `Assembler`.
     pub fn new() -> Self {
         Self::default()
     }
 
+    /// Indicate that the input sequence is complete. Returns any errors which
```
Comment (lightclient) [inline etk-asm/src/asm.rs:182]:
```suggestion
    /// Indicate that the input sequence is complete. Returns any errors that
```

File: etk-asm/src/asm.rs [hunk 3]
```diff
@@ -135,10 +199,14 @@ impl Assembler {
         Ok(())
     }
 
+    /// Collect any assembled instructions which are ready to be output.
```
Comment (lightclient) [inline etk-asm/src/asm.rs:202]:
```suggestion
    /// Collect any assembled instructions that are ready to be output.
```

File: etk-asm/src/asm.rs [hunk 4]
```diff
@@ -135,10 +199,14 @@ impl Assembler {
...
     pub fn take(&mut self) -> Vec<u8> {
         std::mem::take(&mut self.ready)
     }
 
+    /// Feed instructions into the `Assembler`.
+    ///
+    /// Returns the number of bytes which can be collected with [`Assembler::take`].
```
Comment (lightclient) [inline etk-asm/src/asm.rs:209]:
```suggestion
    /// Returns the number of bytes that can be collected with [`Assembler::take`].
```

File: etk-asm/src/asm.rs [hunk 5]
```diff
@@ -153,6 +221,9 @@ impl Assembler {
         Ok(self.ready.len())
     }
 
+    /// Feed a single instruction into the `Assembler`.
+    ///
+    /// Returns the number of bytes which can be collected with [`Assembler::take`]
```
Comment (lightclient) [inline etk-asm/src/asm.rs:226]:
```suggestion
    /// Returns the number of bytes that can be collected with [`Assembler::take`]
```

File: etk-asm/src/disasm.rs [hunk 6]
```diff
@@ -1,21 +1,52 @@
...
+//!
+//! Converts a stream of bytes into an iterator of [`ConcreteOp'].
+//!
+//! See the documentation for [`Disassembler`] for more information.
+mod error {
+    use snafu::{Backtrace, Snafu};
+
+    use super::Offset;
+
+    /// Errors which may arise during disassembly.
```
Comment (lightclient) [inline etk-asm/src/disasm.rs:11]:
```suggestion
    /// Errors that may arise during disassembly.
```

File: etk-asm/src/ingest.rs [hunk 7]
```diff
@@ -1,42 +1,90 @@
...
 mod error {
     use crate::asm::Error as AssembleError;
-    use crate::parse::ParseError;
+    use crate::ParseError;
 
     use snafu::{Backtrace, Snafu};
 
     use std::path::PathBuf;
 
+    /// Errors which may arise during the assembly process.
```
Comment (lightclient) [inline etk-asm/src/ingest.rs:12]:
```suggestion
    /// Errors that may arise during the assembly process.
```

File: etk-asm/src/ingest.rs [hunk 8]
```diff
@@ -1,42 +1,90 @@
...
+            "`{}` is outside of the root directory `{}`",
+            file.display(),
+            root.display()
+        ))]
+        #[non_exhaustive]
         DirectoryTraversal {
+            /// The root directory.
             root: PathBuf,
+
+            /// The file which was to be included or imported.
```
Comment (lightclient) [inline etk-asm/src/ingest.rs:28]:
```suggestion
            /// The file that was to be included or imported.
```

File: etk-asm/src/ingest.rs [hunk 9]
```diff
@@ -1,42 +1,90 @@
...
             source: std::io::Error,
+
+            /// The location of the error.
             backtrace: Backtrace,
+
+            /// The optional path where the error occurred.
             path: Option<PathBuf>,
         },
+
+        /// An error which occurred while parsing a file.
```
Comment (lightclient) [inline etk-asm/src/ingest.rs:49]:
```suggestion
        /// An error that occurred while parsing a file.
```

File: etk-asm/src/ingest.rs [hunk 10]
```diff
@@ -1,42 +1,90 @@
...
         #[snafu(context(false))]
+        #[non_exhaustive]
+        #[snafu(display("{}", source))]
         Parse {
+            /// The underlying source of this error.
             #[snafu(backtrace)]
             source: ParseError,
         },
+
+        /// An error which occurred while assembling a file.
```
Comment (lightclient) [inline etk-asm/src/ingest.rs:59]:
```suggestion
        /// An error that occurred while assembling a file.
```

File: etk-asm/src/ingest.rs [hunk 11]
```diff
@@ -271,12 +319,38 @@ where
     }
 }
 
+/// A high-level interface for assembling file into EVM bytecode.
```
Comment (lightclient) [inline etk-asm/src/ingest.rs:322]:
```suggestion
/// A high-level interface for assembling a file into EVM bytecode.
```

File: etk-asm/src/ops.rs [hunk 12]
```diff
@@ -1,6 +1,23 @@
+//! Definitions of all of the instructions supported by the assembler.
+
+mod error {
+    use snafu::{Backtrace, Snafu};
+
+    /// The error which can arise while parsing a specifier from a string.
```
Comment (lightclient) [inline etk-asm/src/ops.rs:6]:
```suggestion
    /// The error that can arise while parsing a specifier from a string.
```

File: etk-asm/src/ops.rs [hunk 13]
```diff
@@ -261,11 +298,16 @@ macro_rules! ops {
...
             }
 
+            /// Return the total encoded size for this instruction, including the
+            /// immediate if one is required.
             pub fn size(&self) -> u32 {
                 self.specifier().extra_len() + 1u32
             }
         }
 
+        /// A `Specifier` is an `Op` which takes no arguments. It's useful to refer
```
Comment (lightclient) [inline etk-asm/src/ops.rs:308]:
```suggestion
        /// A `Specifier` is an `Op` that takes no arguments. It's useful to refer
```

File: etk-asm/src/ops/imm.rs [hunk 14]
```diff
@@ -5,12 +5,14 @@ use snafu::{Backtrace, Snafu};
 use std::convert::TryFrom;
 use std::fmt::{self, Debug};
 
+/// An error which arises when converting an integer into an immediate.
```
Comment (lightclient) [inline etk-asm/src/ops/imm.rs:8]:
```suggestion
/// An error that arises when converting an integer into an immediate.
```

File: etk-asm/src/ops/imm.rs [hunk 15]
```diff
@@ -5,12 +5,14 @@ use snafu::{Backtrace, Snafu};
...
 #[derive(Snafu, Debug)]
 #[snafu(visibility = "pub(super)")]
 pub struct TryFromIntError {
     backtrace: Backtrace,
 }
 
+/// An error which arises when converting a slice into an immediate.
```
Comment (lightclient) [inline etk-asm/src/ops/imm.rs:15]:
```suggestion
/// An error that arises when converting a slice into an immediate.
```

File: etk-asm/src/parse/error.rs [hunk 16]
```diff
@@ -4,34 +4,62 @@ use snafu::{Backtrace, IntoError, Snafu};
 
 use super::Rule;
 
+/// Type for errors which may arise while parsing assembly source code.
```
Comment (lightclient) [inline etk-asm/src/parse/error.rs:7]:
```suggestion
/// Type for errors that may arise while parsing assembly source code.
```

File: etk-cli/src/io.rs [hunk 17]
```diff
@@ -34,6 +38,7 @@ pub struct InputSource {
 }
 
 impl InputSource {
+    /// Convert `self` into something which implements `std::io::Read`.
```
Comment (lightclient) [inline etk-cli/src/io.rs:41]:
```suggestion
    /// Convert `self` into something that implements `std::io::Read`.
```

File: etk-cli/src/io.rs [hunk 18]
```diff
@@ -58,12 +63,16 @@ impl InputSource {
     }
 }
 
+/// An implementation of `std::io::Write` which converts from binary to
```
Comment (lightclient) [inline etk-cli/src/io.rs:66]:
```suggestion
/// An implementation of `std::io::Write` that converts from binary to
```

**Review by lightclient — APPROVED** [review]
I didn't realize there going to be so many "which"s otherwise, probably wouldn't have started commenting :) I think "that" is generally more correct though since "which" is used more when the phrase following is more optional to the sentence.

File: etk-asm/src/ops.rs [hunk 19]
```diff
@@ -1,6 +1,23 @@
+//! Definitions of all of the instructions supported by the assembler.
```
[the remark this answers is not in the document]
Comment (SamWilsn) [reply etk-asm/src/ops.rs:1]: This one is just being overly picky: https://getproofed.com/writing-tips/word-choice-all-of/

**Pull request merged**